portability-core/src/main/java/org/dataportabilityproject/shared/CloudAppCredentialFactory.java

```diff
@@ -38,24 +55,59 @@
       + "keyRings/portability_secrets/cryptoKeys/portability_secrets_key";
 
   private final CloudFactory cloudFactory;
+  private final BucketStore bucketStore;
+  private final CryptoKeyManagementSystem cryptoKeyManagementSystem;
 
   @Inject
   CloudAppCredentialFactory(CloudFactory cloudFactory) {
     this.cloudFactory = cloudFactory;
+    this.bucketStore = cloudFactory.getBucketStore();
+    this.cryptoKeyManagementSystem = cloudFactory.getCryptoKeyManagementSystem();
+    this.keys = CacheBuilder.newBuilder().expireAfterWrite(10, TimeUnit.MINUTES).build(
+        new CacheLoader<String, String>() {
+          @Override
+          public String load(String key) throws Exception {
+            return lookupKey(key);
+          }
+        });
+    this.secrets = CacheBuilder.newBuilder().expireAfterWrite(10, TimeUnit.MINUTES).build(
+        new CacheLoader<String, String>() {
+          @Override
+          public String load(String key) throws Exception {
+            return lookupSecret(key);
+          }
+        });
   }
 
   @Override
   public AppCredentials lookupAndCreate(String keyName, String secretName) throws IOException {
+    String key;
+    String secret;
+    try {
+      key = keys.get(keyName);
```

Comment:
Could we do each of these in it's own try block so we can get a better exception?

Reply:
Done